### ISSUE TITLE
Deprecate record access by attr in CUDA target

### ISSUE THREAD

**Issue by gmarkall** [numba/numba#5375]
In the CUDA target it is possible to access record items as attrs, for example:

```python
point_type = [('x', np.float64), ('y', np.float64)]
points = np.recarray(1, dtype=point_type)

@cuda.jit
def move_to_origin(point):
    point.x = 0
    point.y = 0

move_to_origin[1, 1](points[0])
```

This is not legal in normal Python / NumPy - removing the `@cuda.jit` would require the following syntactic change in `move_to_origin`:

```python
def move_to_origin(point):
    point['x'] = 0
    point['y'] = 0
```

The former syntax should be deprecated, for consistency with non-cuda-jitted Python code. Any code impacted by the deprecation can be fixed up by substituting the former syntax for the latter.

**Comment by c200chromebook** [comment]
Would recommend throwing a warning for a few releases before turning it off. This will be breaking for a lot of people.

**Comment by stuartarchibald** [comment]
Thanks @gmarkall, I agree with deprecation, @sklam any view to the contrary?

**Comment by stuartarchibald** [comment]
> Would recommend throwing a warning for a few releases before turning it off. This will be breaking for a lot of people.

Thanks :) Numba issues both `PendingDeprecationWarnings` and `DeprecationWarnings` depending on the perceived prevalence of use.

**Comment by sklam** [comment]
Yes, we should warn about it. Also, I think any change to the record access will affect every target. 

IIRC, the history for this behavior mismatch is because `getattr` would have the attribute name in the bytecode. That simplified things at the time before we have any constant propagation.

**Comment by gmarkall** [comment]
Indeed, it does affect every target - I was mistakenly under the impression that this was a CUDA-specific behaviour from when I added record arrays to the CUDA target. It seems that Numba's typing of an `ndarray` with structured type, and a `recarray`, are the same, so it's difficult to distinguish when a deviation from NumPy semantics is actually happening - NumPy allows get/setattr on record arrays but not arrays of structured type. 

I think detecting this to add a deprecation warning would involve adding a new type or two (perhaps a `RecArray` type and a `void` type for the elements of structured arrays) and additional logic to make sure that typing of get/setattr doesn't happen for elements of structured arrays... I'm questioning whether this work is worth the effort, and I'm thinking it might be better just to document this deviation from NumPy.

**Comment by gmarkall** [comment]
Also, this is a duplicate of #878.

**Comment by gmarkall** [comment]
Reopening as I think Github got confused and auto-closed it.

**Comment by cshenton** [comment]
Having this behaviour available is super desirable. If it is removed in future, it would be great if there was an equivalent for cuda arrays to `x.view(np.recarray)` in numpy, so that the deprecation can be fixed by sprinkling that around the existing code which depends on this.
